Sources/Swift/Integrations/SessionReplay/SentryOnDemandReplay.swift

```diff
@@ -135,164 +146,264 @@ class SentryOnDemandReplay: NSObject, SentryReplayVideoMaker {
         var frameCount = 0
-        
+
         var videos = [SentryVideoInfo]()
-        
+
         while frameCount < videoFrames.count {
-            let outputFileURL = URL(fileURLWithPath: _outputPath.appending("/\(videoFrames[frameCount].time.timeIntervalSinceReferenceDate).mp4"))
-            SentryLog.debug("[Session Replay] Rendering video with output file URL: \(outputFileURL)")
-            if let videoInfo = try renderVideo(with: videoFrames, from: &frameCount, at: outputFileURL) {
-                videos.append(videoInfo)
-            } else {
-                frameCount++
+            let frame = videoFrames[frameCount]
+            let outputFileURL = URL(fileURLWithPath: _outputPath)
+                .appendingPathComponent("\(frame.time.timeIntervalSinceReferenceDate)")
+                .appendingPathExtension("mp4")
+
+            let group = DispatchGroup()
+            var currentError: Error?
+
+            group.enter()
+            self.renderVideo(with: videoFrames, from: frameCount, at: outputFileURL) { result in
+                switch result {
+                case .success(let videoResult):
+                    // Set the frame count/offset to the new index that is returned by the completion block.
+                    // This is important to avoid processing the same frame multiple times.
+                    frameCount = videoResult.finalFrameIndex
+                    SentryLog.debug("[Session Replay] Finished rendering video, frame count moved to: \(frameCount)")
+
+                    // Append the video info to the videos array.
+                    // In case no video info is returned, skip the segment.
+                    if let videoInfo = videoResult.info {
+                        videos.append(videoInfo)
+                    }
+                case .failure(let error):
+                    SentryLog.error("[Session Replay] Failed to render video with error: \(error)")
+                    currentError = error
+                }
+                group.leave()
+            }
+
+            // Calling group.wait will block the `processingQueue` until the video rendering completes or a timeout occurs.
+            // It is imporant that the renderVideo completion block signals the group.
+            // The queue used by render video must have a higher priority than the processing queue to reduce thread inversion.
+            // Otherwise, it could lead to queue starvation and a deadlock/timeout.
+            guard group.wait(timeout: .now() + 10) == .success else {
+                SentryLog.error("[Session Replay] Timeout while waiting for video rendering to finish, returning \(videos.count) videos")
+                return videos
             }
+
+            // If there was an error, log it and exit the loop.
+            if let error = currentError {
+                // In previous implementations the error was propagated to the completion block, discarding any generated video.
```

Comment:
`l`: Mentioning a specific SDK version makes this a bit more specific, because if I want to find out which one, I need to do some git blaming.